### PR TITLE
Add is_subset handler for Range in FiniteSet

### LINKED ISSUE #18124: Regression in `is_subset` for Range, FiniteSet
I'm guessing this was introduced in #17848 .

```
import sympy
from sympy.sets import Range, FiniteSet

print(f"SymPy version: {sympy.__version__}")
print(Range(-3, 4, 1).is_subset(FiniteSet(-10, 10)))
```

Output at commit 99663a6d90 on master branch:
```
SymPy version: 1.6.dev
None
```
Behavior is as expected on 1.5:
```
SymPy version: 1.5
False
```

### LINKED ISSUE #18146: Range(1).intersect(FiniteSet(n)) raises TypeError: cannot determine truth value of Relational
```
n = Symbol('n', integer=True)
Range(1).intersect(FiniteSet(n))
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-66-74dcb9ca2d9f> in <module>
----> 1 Range(1).intersect(FiniteSet(n))

/opt/tljh/user/lib/python3.6/site-packages/sympy/sets/sets.py in intersect(self, other)
    138 
    139         """
--> 140         return Intersection(self, other)
    141 
    142     def intersection(self, other):

/opt/tljh/user/lib/python3.6/site-packages/sympy/sets/sets.py in __new__(cls, *args, **kwargs)
   1310         if evaluate:
   1311             args = list(cls._new_args_filter(args))
-> 1312             return simplify_intersection(args)
   1313 
   1314         args = list(ordered(args, Set._infimum_key))

/opt/tljh/user/lib/python3.6/site-packages/sympy/sets/sets.py in simplify_intersection(args)
   2176 
   2177     # Handle Finite sets
-> 2178     rv = Intersection._handle_finite_sets(args)
   2179 
   2180     if rv is not None:

/opt/tljh/user/lib/python3.6/site-packages/sympy/sets/sets.py in _handle_finite_sets(args)
   1395         definite = set()
   1396         for e in all_elements:
-> 1397             inall = fuzzy_and(s.contains(e) for s in args)
   1398             if inall is True:
   1399                 definite.add(e)

/opt/tljh/user/lib/python3.6/site-packages/sympy/core/logic.py in fuzzy_and(args)
    137 
    138     rv = True
--> 139     for ai in args:
    140         ai = fuzzy_bool(ai)
    141         if ai is False:

/opt/tljh/user/lib/python3.6/site-packages/sympy/sets/sets.py in <genexpr>(.0)
   1395         definite = set()
   1396         for e in all_elements:
-> 1397             inall = fuzzy_and(s.contains(e) for s in args)
   1398             if inall is True:
   1399                 definite.add(e)

/opt/tljh/user/lib/python3.6/site-packages/sympy/sets/sets.py in contains(self, other)
    332         """
    333         other = sympify(other, strict=True)
--> 334         c = self._contains(other)
    335         if c is None:
    336             return Contains(other, self, evaluate=False)

/opt/tljh/user/lib/python3.6/site-packages/sympy/sets/fancysets.py in _contains(self, other)
    668         if (ref - other) % self.step:  # off sequence
    669             return S.false
--> 670         return _sympify(other >= self.inf and other <= self.sup)
    671 
    672     def __iter__(self):

/opt/tljh/user/lib/python3.6/site-packages/sympy/core/relational.py in __nonzero__(self)
    374 
    375     def __nonzero__(self):
--> 376         raise TypeError("cannot determine truth value of Relational")
    377 
    378     __bool__ = __nonzero__

TypeError: cannot determine truth value of Relational
```

### PR DESCRIPTION
**Pull request by gschintgen**
#### References to other Issues or PRs

Fixes #18124.
Fixes #18146.

#### Brief description of what is fixed or changed
Add `is_subset` handler for (Range, FiniteSet) and tests.
Fixes logic errors in `Range.contains()`

#### Other comments
Two missed lines (marked red in codecov.io) in sets/handlers/issubset.py should now also be covered.

#### Release Notes

* sets
  * Added an `is_subset` handler to check whether a `Range` is included in a `FiniteSet`.
  * Added a simple cache to `FiniteSet` in order to speed up some membership checks.
  * Fixed `Range.contains()` logic for symbolic values.

### CODE REVIEW

File: sympy/sets/handlers/issubset.py [hunk 1]
```diff
@@ -66,6 +66,18 @@ def is_subset_sets(a_range, b_interval): # noqa:F811
...
+def is_subset_sets(a_range, b_finiteset): # noqa:F811
+    try:
+        a_size = a_range.size
+    except ValueError:
+        # symbolic Range of unknown size
+        return None
+    if a_size > len(b_finiteset):
+        return False
+    else:
+        return fuzzy_and(b_finiteset.contains(x) for x in a_range)
```
Comment (oscarbenjamin) [inline sympy/sets/handlers/issubset.py]: I guess we could use this logic for any finite iterable set rather than just `Range`. It could be slow though e.g.:
```
In [5]: S = FiniteSet(*(Dummy() for _ in range(10**3)))

In [6]: Range(10**3).is_subset(S) # Really slow...
```

Comment (czgdp1807) [reply sympy/sets/handlers/issubset.py]: May be this can be analysed just by using endpoints of `FiniteSet` and `Range`. @gschintgen Have you given this idea a consideration?

Comment (gschintgen) [reply sympy/sets/handlers/issubset.py]: > Range(10**3).is_subset(S) # Really slow...

That's a pity. Even a *single* `S.contains(5)` takes almost 2 seconds (on a slow system). I concede that I must have imagined `FiniteSet` to be hash-based so that a contains-check would be almost instant (and constant-time relative to the cardinality). But in hindsight it's obvious that, in order to give mathematically meaningful answers, `contains` must indeed check equality with every element of the `FiniteSet`. So it comes down to 2ms per equality check which sums up rather quickly. Hmm. (2ms is still expensive for a rather basic check, isn't it?)
So the problem with the proposed code is that its running time scales with the product of the cardinalities of both sets and that's not really acceptable.

@czgdp1807 : I'm not sure I completely understand your idea. A FiniteSet may contain an arbitrary number of arbitrary expressions. What could potentially be done is somehow filtering/pre-processing the FiniteSet in a single run-through (e.g. kick out irrelevant non-integers). I'll think about it.

Comment (gschintgen) [reply sympy/sets/handlers/issubset.py]: Oh well, this took me some time. I've added a new commit that changes the logic. I just hope I got my testing right. Please review, but note the gotcha below that I discovered only very late...

With this new commit, the example given by @oscarbenjamin is handled almost instantly by the handler, **but** there's another **huge** slowdown in line 385:
https://github.com/sympy/sympy/blob/f7390e5fa2313be1c49959270358b2b0c62a1a57/sympy/sets/sets.py#L377-L386
The `None` case is already pathologically slow in current master due to the intersection computation.

I also note that the logic of that intersection test is bound to fail for Range.intersect(FiniteSet): The result will be a FiniteSet and hence test unequal to self. So the expensive computation is in vain.

So to sum up (for the slow example):
master: slow due to useless intersection computation
first commit: slow handler, followed by slow intersection
new commit: faster handler, followed by slow intersection.

Comment (oscarbenjamin) [reply sympy/sets/handlers/issubset.py]: Not sure if I added that comment but it sounds like something I'd say...

Comment (oscarbenjamin) [reply sympy/sets/handlers/issubset.py]: I wonder if it's possible to speed up general membership testing with something like this. Maybe the `FiniteSet` could cache a `set` that can be used for fast membership testing as separate from the more complicated symbolic cases.

Comment (gschintgen) [reply sympy/sets/handlers/issubset.py]: I've added a basic cache. (In fact a temporary `set` of all elements/args was created anyway in `__new__`, so it was sufficient to keep it around.) For larger `FiniteSets` it's indeed beneficial to first check membership using this cache. But only the `True` case with an exact match is sped up.
(See commit message for a cursory benchmark.)

Comment (oscarbenjamin) [reply sympy/sets/handlers/issubset.py]: It doesn't need to be done in this PR but I guess there are some situations where we know based on type that `==` is sufficient and therefore can use efficient set algorithms. For example if all elements of both sets are of type `Number` then we can process everything using `set` because we know that `Eq(a, b)` is True if `a==b` and False otherwise (not sure about `Float`...). On other hand if any element is a symbol with no defined assumptions then we know that `Eq(a, b)` will be True if `a == b` and "None" otherwise so probably something can be done with that as well.

Comment (oscarbenjamin) [reply sympy/sets/handlers/issubset.py]: I think we've probably over-complicated things here. Presumably the simple contains loop would cover the case
```
Range(n, n + 2).is_subset(FiniteSet(n, n + 1))
```
which with the PR now is None rather than True.

I guess we should at least fall back on the slow method here. It's probably more important to do the right thing with small symbolic sets than worry about performance with larger sets.

Comment (gschintgen) [reply sympy/sets/handlers/issubset.py]: > ```
> Range(n, n + 2).is_subset(FiniteSet(n, n + 1))
> ```
> 
> 
> which with the PR now is None rather than True.

I agree that for a CAS symbolics should be more important than pure performance. I'm somewhat astonished that in the early commit with the basic `fuzzy_and` check I included this test with `None` as expected result. (The code of the very first commit indeed returns `True`!) I know I did some trials with symbolic `Ranges` but must have misinterpreted something. Possibly a mishap with the `n` in the tests file that doesn't have integer assumptions combined with #18146 cropping up. I think I even concluded that `Range` and symbols don't mix well...

I'll see what I can do.

Comment (gschintgen) [reply sympy/sets/handlers/issubset.py]: @oscarbenjamin, @czgdp1807 : What do you think now with the latest commits?

As for the more advanced caching, it's an interesting idea and I'm sure there's more that could be done (e.g. not only check if all elements are of certain types, but maybe cache precisely which `args` are of other/symbolic types, so that only those specific elements need to be checked.) But at this stage I'm not sure how important this optimization for large `FiniteSets` is. I'd rather weed out logic errors, sets-related or not.

Comment (oscarbenjamin) [reply sympy/sets/handlers/issubset.py]: > I'd rather weed out logic errors, sets-related or not.

Agreed

**Comment by oscarbenjamin** [comment]
Looks good. Merging.
